Ford: use platform codes to fuzzy fingerprint

incognitojam opened the pull request:
Use the FW prefix to identify the model and model year/generation of the vehicle.

Notes:
- This ignores the "software version" suffix of the FW since we can identify changes in API (e.g. Ford Q3 -> Q4) from the rest of the FW. For example, the 2024 Escape is Q4 and its FW has a new model year hint (see [FW table](https://docs.google.com/spreadsheets/d/1tp6ZYW9fYdDu_xmZWH2Kt39Aw592Ep_7cNJ2kUdPt7M/edit))
- The FW version cannot tell us whether the car is compatible (has LCA/ACC), except the lack of a radar can be used to rule cars out. Some cars have radar without ACC, and some with ACC do not have LCA.
- We can identify the model from the 'platform hint' and model year/generation from the 'model year hint' without any false matches/collisions (see [notebook](https://github.com/incognitojam/op-notebooks/blob/main/projects/all_ecus_platform_codes.ipynb) using big vin database)

**Verification**
- [x] [Checked all segments in commaCarSegments](https://github.com/commaai/openpilot/pull/31124/files#diff-afcb023af37e71f90c90bca319412a3ec9a048241e5178dd762ac75de3548919) to make sure they still fingerprint as expected. The only ones which do not are from a Maverick which is missing a radar anyway.

**Future improvements**
- Remove requirement for all four ECUs to match: we can guarantee a match/no false positives from just three for some platforms

May close #31052

<details>
<summary>Platform codes (platform + model year hints)</summary>

<pre>
FORD BRONCO SPORT 1ST GEN
  (Ecu.eps, 0x730, None):
    Codes: [(b'X6C', b'L')]
  (Ecu.abs, 0x760, None):
    Codes: [(b'X6C', b'L')]
  (Ecu.fwdRadar, 0x764, None):
    Codes: [(b'B5T', b'L')]
  (Ecu.fwdCamera, 0x706, None):
    Codes: [(b'1PT', b'M')]

FORD ESCAPE 4TH GEN
  (Ecu.eps, 0x730, None):
    Codes: [(b'X6C', b'L')]
  (Ecu.abs, 0x760, None):
    Codes: [(b'X6C', b'L')]
  (Ecu.fwdRadar, 0x764, None):
    Codes: [(b'B5T', b'L')]
  (Ecu.fwdCamera, 0x706, None):
    Codes: [(b'J6T', b'L'), (b'V4T', b'L')]

FORD EXPLORER 6TH GEN
  (Ecu.eps, 0x730, None):
    Codes: [(b'1MC', b'L'), (b'1MC', b'M'), (b'1MC', b'P')]
  (Ecu.abs, 0x760, None):
    Codes: [(b'1MC', b'L')]
  (Ecu.fwdRadar, 0x764, None):
    Codes: [(b'B5T', b'L')]
  (Ecu.fwdCamera, 0x706, None):
    Codes: [(b'B5T', b'L'), (b'C5T', b'L')]

FORD F-150 14TH GEN
  (Ecu.eps, 0x730, None):
    Codes: [(b'L3V', b'M')]
  (Ecu.abs, 0x760, None):
    Codes: [(b'L34', b'P')]
  (Ecu.fwdRadar, 0x764, None):
    Codes: [(b'L3T', b'M')]
  (Ecu.fwdCamera, 0x706, None):
    Codes: [(b'J6T', b'P')]

FORD F-150 LIGHTNING 1ST GEN
  (Ecu.abs, 0x760, None):
    Codes: [(b'L38', b'P')]
  (Ecu.fwdRadar, 0x764, None):
    Codes: [(b'L3T', b'M')]
  (Ecu.fwdCamera, 0x706, None):
    Codes: [(b'L3T', b'M')]

FORD MUSTANG MACH-E 1ST GEN
  (Ecu.eps, 0x730, None):
    Codes: [(b'J9C', b'L')]
  (Ecu.abs, 0x760, None):
    Codes: [(b'K9C', b'L')]
  (Ecu.fwdRadar, 0x764, None):
    Codes: [(b'L3T', b'M')]
  (Ecu.fwdCamera, 0x706, None):
    Codes: [(b'L3T', b'M')]

FORD FOCUS 4TH GEN
  (Ecu.eps, 0x730, None):
    Codes: [(b'X6C', b'J')]
  (Ecu.abs, 0x760, None):
    Codes: [(b'X61', b'J')]
  (Ecu.fwdRadar, 0x764, None):
    Codes: [(b'X7T', b'J')]
  (Ecu.fwdCamera, 0x706, None):
    Codes: [(b'X7T', b'J')]

FORD MAVERICK 1ST GEN
  (Ecu.eps, 0x730, None):
    Codes: [(b'Z6C', b'N')]
  (Ecu.abs, 0x760, None):
    Codes: [(b'Z6C', b'N'), (b'Z6C', b'P')]
  (Ecu.fwdRadar, 0x764, None):
    Codes: [(b'Z6T', b'N')]
  (Ecu.fwdCamera, 0x706, None):
    Codes: [(b'Z6T', b'N')]


Car models vs. platform codes:
  (Ecu.eps, 0x730, None):
    (b'X6C', b'L'): ['FORD BRONCO SPORT 1ST GEN', 'FORD ESCAPE 4TH GEN']
    (b'1MC', b'P'): ['FORD EXPLORER 6TH GEN']
    (b'1MC', b'M'): ['FORD EXPLORER 6TH GEN']
    (b'1MC', b'L'): ['FORD EXPLORER 6TH GEN']
    (b'L3V', b'M'): ['FORD F-150 14TH GEN']
    (b'J9C', b'L'): ['FORD MUSTANG MACH-E 1ST GEN']
    (b'X6C', b'J'): ['FORD FOCUS 4TH GEN']
    (b'Z6C', b'N'): ['FORD MAVERICK 1ST GEN']
  (Ecu.abs, 0x760, None):
    (b'X6C', b'L'): ['FORD BRONCO SPORT 1ST GEN', 'FORD ESCAPE 4TH GEN']
    (b'1MC', b'L'): ['FORD EXPLORER 6TH GEN']
    (b'L34', b'P'): ['FORD F-150 14TH GEN']
    (b'L38', b'P'): ['FORD F-150 LIGHTNING 1ST GEN']
    (b'K9C', b'L'): ['FORD MUSTANG MACH-E 1ST GEN']
    (b'X61', b'J'): ['FORD FOCUS 4TH GEN']
    (b'Z6C', b'P'): ['FORD MAVERICK 1ST GEN']
    (b'Z6C', b'N'): ['FORD MAVERICK 1ST GEN']
  (Ecu.fwdRadar, 0x764, None):
    (b'B5T', b'L'): ['FORD BRONCO SPORT 1ST GEN', 'FORD ESCAPE 4TH GEN', 'FORD EXPLORER 6TH GEN']
    (b'L3T', b'M'): ['FORD F-150 14TH GEN', 'FORD F-150 LIGHTNING 1ST GEN', 'FORD MUSTANG MACH-E 1ST GEN']
    (b'X7T', b'J'): ['FORD FOCUS 4TH GEN']
    (b'Z6T', b'N'): ['FORD MAVERICK 1ST GEN']
  (Ecu.fwdCamera, 0x706, None):
    (b'1PT', b'M'): ['FORD BRONCO SPORT 1ST GEN']
    (b'J6T', b'L'): ['FORD ESCAPE 4TH GEN']
    (b'V4T', b'L'): ['FORD ESCAPE 4TH GEN']
    (b'B5T', b'L'): ['FORD EXPLORER 6TH GEN']
    (b'C5T', b'L'): ['FORD EXPLORER 6TH GEN']
    (b'J6T', b'P'): ['FORD F-150 14TH GEN']
    (b'L3T', b'M'): ['FORD F-150 LIGHTNING 1ST GEN', 'FORD MUSTANG MACH-E 1ST GEN']
    (b'X7T', b'J'): ['FORD FOCUS 4TH GEN']
    (b'Z6T', b'N'): ['FORD MAVERICK 1ST GEN']
</pre>
</detail>

<!--- ***** Template: Car Bugfix *****

**Description**

A description of the bug and the fix. Also link the issue if it exists. 

**Verification**

Explain how you tested this bug fix. 

**Route**

Route: [a route with the bug fix]


-->

incognitojam committed:
Ford: use platform codes to fuzzy fingerprint

TODO: write scripts/tests (print platform codes and version ranges etc.)

May close #31052

incognitojam committed:
get_platform_codes: fix return type

incognitojam committed:
add print_platform_codes.py script

incognitojam committed:
Merge remote-tracking branch 'commaai/master' into ford-platform-codes

incognitojam committed:
print_platform_codes: sort versions

incognitojam committed:
match_fw_to_car_fuzzy: use set comprehension, and fix typo

incognitojam committed:
Ford: add missing Mach-E fw

From the route 83a4e056c7072678/2023-11-13--16-51-33 (which is already
in selfdrive/car/tests/routes.py, added in #30691).

incognitojam committed:
Merge branch 'incognitojam/ford-mach-e-fw' into ford-platform-codes

incognitojam committed:
add ford_fuzzy_fingerprint.ipynb notebook

incognitojam committed:
get_platform_codes: use regex to parse firmware

incognitojam committed:
test_ford: test_platform_codes_fuzzy_fw

incognitojam committed:
test_ford: use get_platform_codes in test_fw_versions

incognitojam committed:
match_fw_to_car_fuzzy: improve comments

incognitojam committed:
test_ford: add test_platform_codes_spot_check

incognitojam committed:
test_ford: add test_match_fw_fuzzy

incognitojam committed:
remove comment from notebook

sshane commented:
great work! I love how this is mostly just a copy of our other fuzzy FP function; it mostly works for all brands. With verification, I made sure that:
- [ ] No dongles ever changed fingerprint (including checking all other brands), running match_fw_to_car
- [ ] Check how many mock dongles gained a fingerprint (match_fw_to_car)
- [ ] Check how many Ford fingerprinting dongles (fuzzy or exact) gained a fingerprint, running match_fw_to_car(allow_exact=False)

Let me know if you have any other verification step ideas

incognitojam commented:
I found some VINs for Ford Explorers online (most of our FW PRs are for those) and used https://motorcraft.com/AsBuilt to download information about them (ECU configuration, part numbers and FW versions). I collected 44 VINs and data.

Here are the results of running the current OP fingerprint vs running with the custom fuzzy matcher:
![changes by vin](https://github.com/commaai/openpilot/assets/4038174/704b34c8-16b6-4003-838a-06b9a5e40a39)
![model year stats](https://github.com/commaai/openpilot/assets/4038174/b57f7311-5485-4250-87a3-79446f1f31fb)
![total stats](https://github.com/commaai/openpilot/assets/4038174/8483f717-4447-45f7-8529-91f6e25f67b4)

Of the 15 newly fingerprinted VINs, 100% were previously failing to fingerprint due to missing Engine FWs. For example, `LB5A-14C204-EAE` was not in the offline database, but there are many versions with the same platform code `B5A` and part number `14C204`, so the custom matcher allows it to match.

Of the 21 VINs which continue to not match, three of these (MY17 and MY19) did not have any camera or radar part numbers/fw. The other 18 did have cameras but did not have radars.

sshane commented:
> Of the 21 VINs which continue to not match, three of these (MY17 and MY19) did not have any camera or radar part numbers/fw. The other 18 did have cameras but did not have radars.

So with a perfect fuzzy matcher, we still expect none of these to match? Ie. they all have LKAS but not ACC built in stock?

incognitojam commented:
> So with a perfect fuzzy matcher, we still expect none of these to match? Ie. they all have LKAS but not ACC as stock?

Yes pretty much.

Looking at the cars which are fingerprinting in more detail, there is a car which shouldn't be compatible which is matching right now, without the custom fuzzy matcher.

For example this 2020 Ford Explorer XLT (`1FMSK8DH1LGB19569`) has this FW:
```
Ecu.eps: [b'L1MC-14D003-AK\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00']
Ecu.abs: [b'L1MC-2D053-BA\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00']
Ecu.fwdCamera: [b'LB5T-14F397-AE\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00']
Ecu.engine: [b'LB5A-14C204-BUJ\x00\x00\x00\x00\x00\x00\x00\x00\x00']
```
Note it doesn't have a radar. It fingerprints anyway, I think because fuzzy matching requires just two matching ECUs.
```
Fingerprinted FORD EXPLORER 6TH GEN using fuzzy match. 2 matching ECUs
```
Since camera/radar/eps are excluded from fuzzy, it's the abs and engine FW matching the database. Both were added in https://github.com/commaai/openpilot/commit/fd2e03decf0a89e1c04e0a8d6aee0bf033d2ec0b.

Looking at that PR, the Explorer (`1FMSK8FH6LGB17698`) does have a radar, and all other FW are identical. So either we have to require the radar FW or use some other query to determine whether it has ACC.
```
Ecu.eps: [b'L1MC-14D003-AK\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00']
Ecu.abs: [b'L1MC-2D053-BA\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00']
Ecu.fwdRadar: [b'LB5T-14D049-AB\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00']
Ecu.fwdCamera: [b'LB5T-14F397-AE\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00']
Ecu.engine: [b'LB5A-14C204-BUJ\x00\x00\x00\x00\x00\x00\x00\x00\x00']
```

sshane commented:
I think we should pick a subset of ECUs to care about, and ignore the rest. This random Bronco does not fingerprint because the engine version is out of range: d1aacbb70e8a9d4b|2024-01-27--09-47-46

sheaduncan commented:
A heads up, alot of us are running this UDS fix https://github.com/commaai/panda/pull/1524 to get CANFD responses.

I'm not sure you've tested this with some Mazda data, but I found shared ECUs.
https://github.com/commaai/panda/pull/1524#issuecomment-1858925102

incognitojam committed:
Merge remote-tracking branch 'commaai/master' into ford-platform-codes

incognitojam committed:
TestFordFW: remove engine ECU FW

incognitojam committed:
update print_platform_codes.py

incognitojam commented:
I've been writing [scripts](https://github.com/incognitojam/op-notebooks/blob/main/ford/ford_fw_research.ipynb) to pull information from the NHTSA and Ford Motorcraft site for a scraped list of VINs so that I can find patterns in the software versions.

For the Explorer (and Aviator), the presence of a radar ECU is enough to confirm that the car has ACC and Lane Centering. For all the other ECUs the software versions are common between cars both with and without ACC and Lane Centering.

Ecu | Part Number | Software | ACC? | Lane Centering?
-- | -- | -- | -- | --
| radar | - | - | No | No
| | LB5T-14F089-AA | LB5T-14D049-AB | Yes | Yes
| camera | - | - | No | No
| | LB5T-14F403-CA | LB5T-14F397-AD | Both | Both
| | | LB5T-14F397-AE | Both | Both
| | | LB5T-14F397-AF | Both | Both
| | LC5T-14F403-FA | LC5T-14F397-AH | Yes | Yes
| eps | GG13-14F079-AD | HB53-14D003-AF | No | No
| | L1MC-14F079-AE | L1MC-14D003-AJ | Both | Both
| | | L1MC-14D003-AK | Both | Both
| | | L1MC-14D003-AL | Both | Both
| | | M1MC-14D003-AB | Yes | Yes
| | | M1MC-14D003-AC | Yes | Yes
| | M1MC-14F079-AA | L1MC-14D003-AL | Both | Both
| | | M1MC-14D003-AB | Both | Both
| | | M1MC-14D003-AC | Both | Both
| | P1MC-14F079-AA | P1MC-14D003-AA | Both | Both
| abs | FB53-14F065-BA | HB53-2D053-BA | No | No
| | L1MC-14F065-AB | L1MC-2D053-AJ | Yes | Yes
| | | L1MC-2D053-BA | Both | Both
| | | L1MC-2D053-BD | Yes | Yes
| | | L1MC-2D053-BF | Yes | Yes
| | | L1MC-2D053-BJ | No | No
| | L1MC-14F065-CB | L1MC-2D053-BB | Both | Both
| | | L1MC-2D053-BC | Both | Both
| | | L1MC-2D053-BD | Both | Both
| | | L1MC-2D053-BE | Both | Both
| | | L1MC-2D053-BF | Both | Both
| | | L1MC-2D053-BJ | Both | Both
| | L1MC-14F065-DA | L1MC-2D053-KB | Both | Both

For the Escape (and Bronco Sport) there is no way of identifying whether the car has both ACC and Lane Centering from the FW response alone. Ford packages ACC and LCA separately on these cars, and the ECU configuration which determines whether LCA is active isn't recorded in the software version.

Soon I'll check if we can query other identifiers from the camera to see whether LCA is enabled. There is likely data on CAN we could use otherwise, like how we check that adaptive cruise is enabled (and not normal cruise). I'm not sure which is preferable.

I'll continue working on collecting the FW versions since it will still be useful for fuzzy fingerprinting.

incognitojam commented:
I collected FW for almost 2000 VINs and produced [a table](https://docs.google.com/spreadsheets/d/1tp6ZYW9fYdDu_xmZWH2Kt39Aw592Ep_7cNJ2kUdPt7M/edit) grouped by the different part numbers/fw versions for the abs, eps, fwdCamera and fwdRadar.

It look as though only the ABS firmware is not shared between models. The PSCM firmware is shared between the F-150 and Mustang Mach-E in a couple of cases. There are many cases of firmware being shared between cars with and without ACC or lane centering enabled.

To determine whether ACC and LCA are enabled we can send queries to the ABS, PSCM and potentially the PCM. If we can't reach the PCM behind the gateway with this query then we can still detect whether the cruise mode is adaptive when onroad from the CAN signal.

incognitojam committed:
Merge remote-tracking branch 'commaai/master' into ford-platform-codes

incognitojam committed:
remove part number (unecessary)

incognitojam committed:
platform codes can just use platform hint and model year - software revision not useful

incognitojam committed:
fuzzy FP on the platform hint and model year hint range

incognitojam committed:
fix platform codes test

incognitojam committed:
update notebook

incognitojam committed:
Merge remote-tracking branch 'commaai/master' into ford-platform-codes

incognitojam committed:
Merge remote-tracking branch 'commaai/master' into ford-platform-codes

incognitojam committed:
Merge remote-tracking branch 'commaai/master' into ford-platform-codes

incognitojam committed:
Merge remote-tracking branch 'commaai/master' into ford-platform-codes

incognitojam commented:
I updated the matching to only look at the "model year" and "platform code" hints in the FW (so, the first four characters) and ignore the software version.

If we considered the software version, we could allow any within the known range of software for that platform/ecu (for example, we could allow `AB` if we had seen `AA` and `AC`), but we normally receive PRs for consecutive FW versions anyway and there aren't many gaps:

<details>
<summary>Old platform codes</summary>
<br/>
<strong>PlatformHint: [ModelYearHint-SoftwareRevision]</strong>

<pre>
FORD BRONCO SPORT 1ST GEN
  (Ecu.eps, 0x730, None):
    X6C: ['L-AAH', 'L-AAK']  <- could allow ['L-AAI', 'L-AAJ']
  (Ecu.abs, 0x760, None):
    X6C: ['L-ARD', 'L-ARE']  <- no new versions!
  (Ecu.fwdRadar, 0x764, None):
    B5T: ['L-AAB']
  (Ecu.fwdCamera, 0x706, None):
    1PT: ['M-AAC']
FORD ESCAPE 4TH GEN
  (Ecu.eps, 0x730, None):
    X6C: ['L-AAF', 'L-AAH', 'L-AAK', 'L-AAL']  <- could allow ['L-AAG', 'L-AAI', 'L-AAJ']
  (Ecu.abs, 0x760, None):
    X6C: ['L-ANS', 'L-ANT', 'L-ANY', 'L-ASA', 'L-ASD']  <- could allow ['L-ANU', 'L-ANX, 'L-ASB', 'L-ASC']
  (Ecu.fwdRadar, 0x764, None):
    B5T: ['L-AAB']
  (Ecu.fwdCamera, 0x706, None):
    J6T: ['L-AAD', 'L-AAE']
    V4T: ['L-AGG']
FORD EXPLORER 6TH GEN
  (Ecu.eps, 0x730, None):
    1MC: ['L-AAJ', 'L-AAK', 'L-AAL', 'M-AAB', 'M-AAC', 'P-AAA']  <- doesn't give us any new versions!
  (Ecu.abs, 0x760, None):
    1MC: ['L-AAJ', 'L-ABA', 'L-ABB', 'L-ABD', 'L-ABF', 'L-AKB']  <- could allow ['L-ABC', 'L-ABE']
  (Ecu.fwdRadar, 0x764, None):
    B5T: ['L-AAB']
  (Ecu.fwdCamera, 0x706, None):
    B5T: ['L-AAD', 'L-AAE', 'L-AAF']
    C5T: ['L-AAE', 'L-AAH']  <- could allow ['L-AAF', 'L-AAG']
FORD F-150 14TH GEN
  (Ecu.eps, 0x730, None):
    L3V: ['M-ABC']
  (Ecu.abs, 0x760, None):
    L34: ['P-ACA']
  (Ecu.fwdRadar, 0x764, None):
    L3T: ['M-AAL']
  (Ecu.fwdCamera, 0x706, None):
    J6T: ['P-ABJ']
FORD F-150 LIGHTNING 1ST GEN
  (Ecu.abs, 0x760, None):
    L38: ['P-AAA']
  (Ecu.fwdRadar, 0x764, None):
    L3T: ['M-AAL']
  (Ecu.fwdCamera, 0x706, None):
    L3T: ['M-ABT']
FORD MUSTANG MACH-E 1ST GEN
  (Ecu.eps, 0x730, None):
    J9C: ['L-AAM', 'L-ACC']
  (Ecu.abs, 0x760, None):
    K9C: ['L-ACK']
  (Ecu.fwdRadar, 0x764, None):
    L3T: ['M-AAL']
  (Ecu.fwdCamera, 0x706, None):
    L3T: ['M-ABS']
FORD FOCUS 4TH GEN
  (Ecu.eps, 0x730, None):
    X6C: ['J-AAH']
  (Ecu.abs, 0x760, None):
    X61: ['J-ACJ']
  (Ecu.fwdRadar, 0x764, None):
    X7T: ['J-AAC']
  (Ecu.fwdCamera, 0x706, None):
    X7T: ['J-AAH']
FORD MAVERICK 1ST GEN
  (Ecu.eps, 0x730, None):
    Z6C: ['N-AAL']
  (Ecu.abs, 0x760, None):
    Z6C: ['N-AAG', 'P-AED', 'P-AEE']
  (Ecu.fwdRadar, 0x764, None):
    Z6T: ['N-AAA']
  (Ecu.fwdCamera, 0x706, None):
    Z6T: ['N-AAC']
</pre>
</details>

incognitojam committed:
add notebook

sshane commented:
I believe we ignore the software versions on Hyundai, but we use the date code to block new model years (sometimes old parts are used on newer cars but it's rare) as Hyundai usually introduces new features/differences (difficulties) when they bump the date code. See: https://github.com/commaai/openpilot/discussions/30810 - we don't want to nor do we fuzzy fingerprint on any of these cars.

We can potentially switch to the software versions for HKG but so far the date code is sufficient to block new cars from fuzzy fingerprinting just because how different new Hyundais are. Since Ford's API is way more static I would think about ways to possibly relax this.

incognitojam committed:
Merge remote-tracking branch 'origin/master' into ford-platform-codes

incognitojam commented:
> Since Ford's API is way more static I would think about ways to possibly relax this if we determine it safe.

I think the only path forward for Ford is to just identify the platform using the FW, and detect LKAS/ACC availability later, since there is no pattern in the FW representing the packages.

We already prevent engagement for non-adaptive cruise, and for most models (except at least the Bronco Sport and Escape) the LKAS and ACC features we require are in the same package (i.e. if ACC is available then LKAS just works). We will be able to catch the missing LKAS API later either using the UDS queries already merged, or finding some CAN signal.

incognitojam committed:
explain model year hint better

incognitojam committed:
test part numbers again

incognitojam committed:
cleanup notebooks

incognitojam committed:
remove notebook

incognitojam committed:
cleanup match_fw_to_car_fuzzy and add comments

incognitojam committed:
update comment

incognitojam committed:
.

incognitojam committed:
Revert "remove notebook"

This reverts commit 5d4ca202f2a23601d5c829204119f36a58f2b451.

incognitojam committed:
add notebook back

incognitojam committed:
remove PSCM from PLATFORM_CODE_ECUS

ABS and IPMA are the best for uniquely matching, and the radar is always required

incognitojam committed:
Revert "remove PSCM from PLATFORM_CODE_ECUS"

This reverts commit b7baeac19c18b5aa0c31da52f12054f4bae6e1ff.

incognitojam committed:
Merge remote-tracking branch 'origin/master' into ford-platform-codes

sshane commented on selfdrive/car/ford/tests/print_platform_codes.py:
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env python3
```
This should be a common script that takes brand eventually

incognitojam committed:
Merge remote-tracking branch 'origin/master' into ford-platform-codes

sshane committed:
Merge remote-tracking branch 'upstream/master' into ford-platform-codes

sshane committed:
fix from merge

sshane committed:
more fixes

revert

sshane committed:
FW_RE -> FW_PATTERN

sshane committed:
this can actually be set

sshane committed:
conventions

sshane committed:
just add

sshane committed:
convention

sshane committed:
refactor matcher, this brings it more in line with Hyundai. IMPORTANT NOTE:

NOTE THAT WE remove the separation for the different platform code model year hint ranges, I don't see that being a problem

sshane committed:
better/smaller test

sshane committed:
add test to catch overlapping platform codes

sshane committed:
remove nb

sshane committed:
Merge remote-tracking branch 'upstream/master' into ford-platform-codes

sshane committed:
not now
